File: Keys and Rooms/keys-and-rooms.py
class Solution(object):
    def canVisitAllRooms(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: bool
        """
        def dfs(rooms, visited, room):
            visited[room] = True
            for key in rooms[room]:
                if not visited[key]:
                    dfs(rooms, visited, key)

        n = len(rooms)
        visited = [False] * n
        dfs(rooms, visited, 0)
        return all(visited)

File: Keys and Rooms/keys-and-rooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int n = rooms.size();
        boolean[] visited = new boolean[n];
        dfs(rooms, visited, 0);
        
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                return false;
            }
        }
        return true;
    }

    private void dfs(List<List<Integer>> rooms, boolean[] visited, int room) {
        visited[room] = true;
        
        for (int key : rooms.get(room)) {
            if (!visited[key]) {
                dfs(rooms, visited, key);
            }
        }
    }
}
